src/torchcodec/_core/BetaCudaDeviceInterface.cpp

```diff
@@ -0,0 +1,563 @@
+  procParams.unpaired_field = dispInfo.repeat_first_field < 0;
+  CUdeviceptr framePtr = 0;
+  unsigned int pitch = 0;
+
+  // We know the frame we want was sent to the hardware decoder, but now we need
+  // to "map" it to an "output surface" before we can use its data. This is a
+  // blocking calls that waits until the frame is fully decoded and ready to be
+  // used.
+  CUresult result = cuvidMapVideoFrame(
+      static_cast<CUvideodecoder>(decoder_.get()),
+      dispInfo.picture_index,
+      &framePtr,
+      &pitch,
+      &procParams);
+
+  if (result != CUDA_SUCCESS) {
+    return AVERROR_EXTERNAL;
+  }
+
+  avFrame = convertCudaFrameToAVFrame(framePtr, pitch, dispInfo);
+
+  // Unmap the frame so that the decoder can reuse its corresponding output
+  // surface. Whether this is blocking is unclear?
+  cuvidUnmapVideoFrame(static_cast<CUvideodecoder>(decoder_.get()), framePtr);
+  // TODONVDEC P0: Get clarity on this:
+  // We assume that the framePtr is still valid after unmapping. That framePtr
+  // is now part of the avFrame, which we'll return to the caller, and the
+  // caller will immediately use it for color-conversion, at which point a copy
+  // happens. After the copy, it doesn't matter whether framePtr is still valid.
+  // And we'll return to this function (and to cuvidUnmapVideoFrame()) *after*
+  // the copy is made, so there should be no risk of overwriting the data before
+  // the copy.
+  // Buuuut yeah, we need get more clarity on what actually happens, and on
+  // what's needed. IIUC DALI makes the color-conversion copy immediately after
+  // cuvidMapVideoFrame() and *before* cuvidUnmapVideoFrame() with a synchronize
+  // in between. So maybe we should do the same.
+
+  return AVSUCCESS;
+}
+
+UniqueAVFrame BetaCudaDeviceInterface::convertCudaFrameToAVFrame(
+    CUdeviceptr framePtr,
+    unsigned int pitch,
+    const CUVIDPARSERDISPINFO& dispInfo) {
+  TORCH_CHECK(framePtr != 0, "Invalid CUDA frame pointer");
+
+  // Get frame dimensions from video format display area (not coded dimensions)
+  // This matches DALI's approach and avoids padding issues
+  int width = videoFormat_.display_area.right - videoFormat_.display_area.left;
+  int height = videoFormat_.display_area.bottom - videoFormat_.display_area.top;
+
+  TORCH_CHECK(width > 0 && height > 0, "Invalid frame dimensions");
+  TORCH_CHECK(
+      pitch >= static_cast<unsigned int>(width), "Pitch must be >= width");
+
+  UniqueAVFrame avFrame(av_frame_alloc());
+  TORCH_CHECK(avFrame.get() != nullptr, "Failed to allocate AVFrame");
+
+  avFrame->width = width;
+  avFrame->height = height;
+  avFrame->format = AV_PIX_FMT_CUDA;
+  avFrame->pts = dispInfo.timestamp; // == guessedPts
+
+  unsigned int frameRateNum = videoFormat_.frame_rate.numerator;
+  unsigned int frameRateDen = videoFormat_.frame_rate.denominator;
+  int64_t duration = static_cast<int64_t>((frameRateDen * timeBase_.den)) /
+      (frameRateNum * timeBase_.num);
```

Comment:
We should probably pull the time utilities out of `SingleStreamDecoder` and put them in `FFMPEGCommon`. It's reasonable to do since they're really time utilities for FFmpeg.

Reply:
Hm, the only uses of `AVRational` I can find in `SingleStreamDecoder` is

```
double ptsToSeconds(int64_t pts, const AVRational& timeBase) {
  // To perform the multiplication before the division, av_q2d is not used
  return static_cast<double>(pts) * timeBase.num / timeBase.den;
}

int64_t secondsToClosestPts(double seconds, const AVRational& timeBase) {
  return static_cast<int64_t>(
      std::round(seconds * timeBase.den / timeBase.num));
}
```


but that's a converting ints to and from floats. Here, we're computing a duration while keeping all pts as `int`.

Reply:
Makes sense - I will add a P0 TODO to move all these into FFmpegCommon. I just want to slightly delay that if that's OK, because that part will slightly change soon when I address the potential zero division issue.